modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/GridCacheLockImpl2Fair.java

```diff
@@ -0,0 +1,657 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.datastructures;
+
+import java.io.Externalizable;
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import javax.cache.processor.EntryProcessorResult;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.IgniteInterruptedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.cache.GridCacheIdMessage;
+import org.apache.ignite.internal.processors.cache.IgniteInternalCache;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteInClosure;
+import org.apache.ignite.plugin.extensions.communication.MessageReader;
+import org.apache.ignite.plugin.extensions.communication.MessageWriter;
+
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.P2P_POOL;
+
+/** Fair implementation of reentrant lock based on IgniteCache.invoke. */
+public final class GridCacheLockImpl2Fair extends GridCacheLockEx2 {
+    /** Reentrant lock name. */
+    private final String name;
+
+    /** Key for shared lock state. */
+    private final GridCacheInternalKey key;
+
+    /** Cache context. */
+    private final GridCacheContext<GridCacheInternalKey, GridCacheLockState2Base<LockOwner>> ctx;
+
+    /** Internal synchronization object. */
+    private final GlobalFairSync sync;
+
+    /** */
+    public GridCacheLockImpl2Fair(String name,
+        GridCacheInternalKey key,
+        IgniteInternalCache<GridCacheInternalKey, GridCacheLockState2Base<LockOwner>> lockView) {
+
+        assert name != null;
+        assert key != null;
+        assert lockView != null;
+
+        this.name = name;
+        this.key = key;
+        this.ctx = lockView.context();
+
+        IgniteLogger log = ctx.logger(getClass());
+
+        sync = new GlobalFairSync(ctx.localNodeId(), key, lockView, ctx, log);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @Override public String name() {
+        return name;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void lock() throws IgniteException {
+        ctx.kernalContext().gateway().readLock();
+
+        try {
+            sync.acquire();
+        }
+        finally {
+            ctx.kernalContext().gateway().readUnlock();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void lockInterruptibly() throws IgniteException {
+        ctx.kernalContext().gateway().readLock();
+
+        try {
+            try {
```

Comment:
[https://ci.memecdn.com/9508525.jpg](url)

Reply:
"we put try into try so you can try and catch everything"

Reply:
fixed